src/main/java/com/direwolf20/buildinggadgets/common/building/implementation/ConnectedSurface.java

```diff
@@ -0,0 +1,123 @@
+import com.google.common.collect.AbstractIterator;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.EnumFacing.Axis;
+import net.minecraft.util.math.BlockPos;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.BitSet;
+import java.util.Iterator;
+import java.util.Queue;
+
+public class ConnectedSurface implements IPlacementSequence {
+
+    public static ConnectedSurface create(BlockPos posHit, EnumFacing sideHit, int range) {
+        return new ConnectedSurface(posHit, sideHit, range);
+    }
+
+    private final Region region;
+    private final BlockPos posHit;
+    private final EnumFacing sideHit;
+
+    protected ConnectedSurface(BlockPos posHit, EnumFacing sideHit, int range) {
+        this.region = Wall.clickedSide(posHit, sideHit, range).getBoundingBox();
+        this.posHit = posHit.toImmutable();
+        this.sideHit = sideHit;
+    }
+
+    /**
+     * For {@link #copy()}
+     */
+    private ConnectedSurface(Region region, BlockPos posHit, EnumFacing sideHit) {
+        this.region = region;
+        this.posHit = posHit;
+        this.sideHit = sideHit;
+    }
+
+    @Override
+    public Region getBoundingBox() {
+        return region;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @return {@code getBoundingBox().contains(x, y, z)} since it would be costly to check each position
+     */
+    @Override
+    public boolean contains(int x, int y, int z) {
+        return region.contains(x, y, z);
+    }
+
+    /**
+     * @deprecated ConnectedSurface should be immutable, so this is not needed
+     */
+    @Deprecated
+    @Override
+    public IPlacementSequence copy() {
+        return new ConnectedSurface(region, posHit, sideHit);
+    }
+
+    /**
+     * @implNote technically BFS fits better with a flood fill algorithm and it simpler to implement with a generator
+     */
+    @Nonnull
+    @Override
+    public Iterator<BlockPos> iterator() {
+        return new AbstractIterator<BlockPos>() {
+            private Queue<BlockPos> queue = new ArrayDeque<>(region.size());
+            private BitSet visited = new BitSet();
+
+            {
+                queue.add(posHit);
+                visited.set(toPlaner(posHit));
+            }
+
+            @Override
+            protected BlockPos computeNext() {
+                if (queue.isEmpty()) {
+                    return endOfData();
+                }
+
+                BlockPos current = queue.remove();
+
+                for (int i = -1; i <= 1; i++) {
+                    for (int j = -1; j <= 1; j++) {
+                        BlockPos neighbor = VectorTools.perpendicularSurfaceOffset(current, sideHit, i, j);
+                        int planer = toPlaner(neighbor);
+
+                        if (visited.get(planer)) {
+                            continue;
+                        }
+                        visited.set(planer);
+
+                        if (region.contains(neighbor)) {
```

Comment:
you might have to move this check to an Sorted Data Structure for Performance (that's what killed the DestructionGadget's Performance, before we fixed it)

Reply:
Sorry, but I don't understand what are you talking about.
`Region#contains` is just several integer comparisons, which I don't think would cause much overhead. Same with `toPlaner(int)` and `BitSet#set`.

Reply:
argh... sorry wrong line

Reply:
Correction I misread: I read "visited.contains"...